File: java/Tutorials/Devices/Java/sound-capture/src/main/java/com/neurotec/tutorials/SoundCapture.java
package com.neurotec.tutorials;

import java.io.IOException;
import java.util.EnumSet;

import com.neurotec.devices.NDevice;
import com.neurotec.devices.NDeviceManager;
import com.neurotec.devices.NDeviceType;
import com.neurotec.devices.NMicrophone;
import com.neurotec.lang.NCore;
import com.neurotec.lang.NThrowable;
import com.neurotec.licensing.NLicense;
import com.neurotec.sound.NSoundBuffer;
import com.neurotec.tutorials.util.LibraryManager;
import com.neurotec.tutorials.util.Utils;

public final class SoundCapture {

	private static final String DESCRIPTION = "Demonstrates sound sample capturing from microphone";
	private static final String NAME = "sound-capture";
	private static final String VERSION = "9.0.0.0";

	private static void Usage() {
		System.out.println("usage:");
		System.out.format("\t{%s [bufferCount]%n", NAME);
		System.out.println();
		System.out.println("\tbufferCount - number of sound buffers to capture from each microphone");
		System.out.println();
	}

	public static void main(String[] args) {
		final String components = "Devices.Microphones";

		LibraryManager.initLibraryPath();

		Utils.printTutorialHeader(DESCRIPTION, NAME, VERSION, args);

		if (args.length < 1) {
			Usage();
			System.exit(1);
		}

		NDeviceManager deviceManager = null;
		try {
			if (!NLicense.obtainComponents("/local", 5000, components)) {
				System.err.format("Could not obtain licenses for components: %s%n", components);
				System.exit(-1);
			}

			int bufferCount = Integer.parseInt(args[0]);
			if (bufferCount == 0) {
				System.err.println("no sound buffers will be captured as sound buffer count is not specified");
				System.exit(-1);
			}
			deviceManager = new NDeviceManager();
			deviceManager.setDeviceTypes(EnumSet.of(NDeviceType.MICROPHONE));
			deviceManager.setAutoPlug(true);
			deviceManager.initialize();
			System.out.format("device manager created. found microphones: %s%n", deviceManager.getDevices().size());

			for (NDevice device : deviceManager.getDevices()) {
				NMicrophone microphone = (NMicrophone) device;
				System.out.format("found microphone %s", microphone.getDisplayName());
				microphone.startCapturing();

				if (bufferCount > 0) {
					System.out.println(", capturing");
					for (int i = 0; i < bufferCount; i++) {
						NSoundBuffer soundSample = microphone.getSoundSample();
						System.out.format("sample buffer received. sample rate: %s, sample length: %s%n", soundSample.getSampleRate(), soundSample.getLength());

						System.out.print(" ... ");
					}
					System.out.print(" done");
					System.out.println();
				}
				microphone.stopCapturing();
			}
			System.out.println("done");
		} catch (Throwable th) {
			Utils.handleError(th);
		} finally {
			if (deviceManager != null) deviceManager.dispose();
			try {
				NLicense.releaseComponents(components);
			} catch (IOException e) {
				e.printStackTrace();
			}
			NCore.shutdown();
		}
	}
}

File: java/Tutorials/Devices/C/SoundCapture/SoundCapture.c
#include <TutorialUtils.h>

#ifdef N_MAC_OSX_FRAMEWORKS
	#include <NCore/NCore.h>
	#include <NMedia/NMedia.h>
	#include <NDevices/NDevices.h>
	#include <NLicensing/NLicensing.h>
#else
	#include <NCore.h>
	#include <NMedia.h>
	#include <NDevices.h>
	#include <NLicensing.h>
#endif

const NChar title[] = N_T("SoundCapture");
const NChar description[] = N_T("Demonstrates capturing sound from microphones");
const NChar version[] = N_T("9.0.0.0");
const NChar copyright[] = N_T("Copyright (C) 2011-2017 Neurotechnology");

int usage()
{
	printf(N_T("usage:\n"));
	printf(N_T("\t%s [bufferCount]\n"), title);
	printf(N_T("\n"));
	printf(N_T("\tbufferCount - number of sound buffers to capture from each microphone to current directory\n"));
	printf(N_T("\n"));
	return 1;
}

int main(int argc, NChar **argv)
{
	HNDeviceManager hDeviceManager = NULL;
	HNDevice hDevice = NULL;
	HNSoundBuffer hSoundSample = NULL;
	HNString hDisplayName = NULL;
	NResult result;
	NInt deviceCount, bufferCount = 0;
	NInt i;
	const NChar * components = N_T("Devices.Microphones");
	NBool available = NFalse;

	OnStart(title, description, version, copyright, argc, argv);

	if (argc < 2)
	{
		OnExit();
		return usage();
	}

	// check the license first
	result = NLicenseObtainComponents(N_T("/local"), N_T("5000"), components, &available);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NLicenseObtainComponents() failed, result = %d\n"), result);
		goto FINALLY;
	}

	if (!available)
	{
		printf(N_T("Licenses for %s not available\n"), components);
		result = N_E_NOT_ACTIVATED;
		goto FINALLY;
	}

	bufferCount = atoi(argv[1]);
	if (bufferCount == 0)
	{
		printf(N_T("no sound buffers will be captured as sound buffer count is not specified"));
	}

	printf(N_T("creating device manager ...\n"));
	result = NDeviceManagerCreateEx(&hDeviceManager);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to create device manager failed (result = %d)\n"), result);
		goto FINALLY;
	}
	result = NDeviceManagerSetDeviceTypes(hDeviceManager, ndtMicrophone);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to set device types (result = %d)!\n"), result);
		goto FINALLY;
	}
	result = NDeviceManagerSetAutoPlug(hDeviceManager, NTrue);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to set auto plug (result = %d)!\n"), result);
		goto FINALLY;
	}

	result = NDeviceManagerInitialize(hDeviceManager);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to initialize device manager (result = %d)!\n"), result);
		goto FINALLY;
	}
	printf(N_T("done\n"));

	/** get number of cameras in the system */
	result = NDeviceManagerGetDeviceCount(hDeviceManager, &deviceCount);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to get number of microphones (result = %d)!\n"), result);
		goto FINALLY;
	}

	/** enumerate cameras */
	for (i = 0; i < deviceCount; i++)
	{
		NInt j;
		const NChar * szDisplayName;

		result = NDeviceManagerGetDevice(hDeviceManager, i, &hDevice);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("failed to get device (result = %d)!\n"), result);
			goto FINALLY;
		}

		/** get camera settings */
		result = NDeviceGetDisplayNameN(hDevice, &hDisplayName);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("failed to get device name (result = %d)!\n"), result);
			continue;
		}

		result = NStringGetBuffer(hDisplayName, NULL, &szDisplayName);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("failed to get device name (result = %d)!\n"), result);
			goto FINALLY;
		}

		printf(N_T("found microphone: %s"), szDisplayName);

		result = NCaptureDeviceStartCapturing(hDevice);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("\nfailed to start microphone (result = %d)!\n"), result);
			goto FINALLY;
		}

		printf(N_T(", capturing:\n"));
		for (j = 0; j < bufferCount; j++)
		{
			NInt sampleRate, length;

			/** get sound buffer from microphone */
			result = NMicrophoneGetSoundSample(hDevice, &hSoundSample);
			if (NFailed(result))
			{
				result = PrintErrorMsgWithLastError(N_T("failed to get sound buffer from microphone (result = %d)!\n"), result);
				goto FINALLY;
			}

			result = NSoundBufferGetSampleRate(hSoundSample, &sampleRate);
			if (NFailed(result))
			{
				result = PrintErrorMsgWithLastError(N_T("failed to get sound buffer sample rate (result = %d)!\n"), result);
				goto FINALLY;
			}

			result = NSoundBufferGetLength(hSoundSample, &length);
			if (NFailed(result))
			{
				result = PrintErrorMsgWithLastError(N_T("failed to get sound buffer sample length (result = %d)!\n"), result);
				goto FINALLY;
			}

			printf(N_T("sample buffer received. sample rate: %d, sample length: %d\n"), sampleRate, length);

			printf(N_T(" ... "));

			result = NObjectSet(NULL, &hSoundSample);
			if (NFailed(result))
			{
				result = PrintErrorMsgWithLastError(N_T("failed to sound buffer (result = %d)!\n"), result);
				goto FINALLY;
			}
		}

		result = NCaptureDeviceStopCapturing(hDevice);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("failed to stop microphone (result = %d)!\n"), result);
			goto FINALLY;
		}

		result = NStringSet(NULL, &hDisplayName);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("failed to clear display name (result = %d)!\n"), result);
			goto FINALLY;
		}

		result = NObjectSet(NULL, &hDevice);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("failed to clear device (result = %d)!\n"), result);
			goto FINALLY;
		}

		printf(N_T(" done\n"));
	}
	printf(N_T("done\n"));

	result = N_OK;
FINALLY:
	{
		NResult result2 = NObjectSet(NULL, &hDevice);
		if (NFailed(result2)) PrintErrorMsg(N_T("NObjectSet() failed (result = %d)!"), result2);
		result2 = NObjectSet(NULL, &hDeviceManager);
		if (NFailed(result2)) PrintErrorMsg(N_T("NObjectSet() failed (result = %d)!"), result2);
		result2 = NObjectSet(NULL, &hSoundSample);
		if (NFailed(result2)) PrintErrorMsg(N_T("NObjectSet() failed (result = %d)!"), result2);
		result2 = NStringSet(NULL, &hDisplayName);
		if (NFailed(result2)) PrintErrorMsg(N_T("NStringSet() failed (result = %d)!"), result2);

		result2 = NLicenseReleaseComponents(components);
		if (NFailed(result2)) PrintErrorMsg(N_T("NLicenseReleaseComponents() failed, result = %d\n"), result2);
	}

	OnExit();

	return result;
}
